4.1 runtime renaming

aholtkamp opened the pull request.

fwienber committed:
[TS] WIP: Adapt to jangaroo-runtime name / import changes

aholtkamp committed:
synced renaming with @jangaroo/runtime refactoring

aholtkamp committed:
added new linter rules: import/no-duplicate, import/order

fwienber commented on jangaroo/jangaroo-compiler/src/main/java/net/jangaroo/jooc/backend/TypeScriptCodeGenerator.java:
```diff
@@ -101,6 +101,41 @@
+          "KeyEvent"
+  );
+  private static final List<String> JANGAROO_RUNTIME_REMOVED_NAMES = Arrays.asList(
+          "AS3.Error",
```
AS3.Error is an adapter that allows Ext JS classes to extend Error, so it is no longer needed in TypeScript. However, it was never directly referenced in ActionScript, but substituted whenever `Error` is used. Thus, there should be no warning.
```suggestion
```

fwienber commented on jangaroo/jangaroo-compiler/src/main/java/net/jangaroo/jooc/backend/TypeScriptCodeGenerator.java:
```diff
@@ -101,6 +101,41 @@
+          "joo.localization",
+          "joo.startTime",
+          "KeyEvent"
+  );
```
I think we can replace this logic by checking all Compilation Units whether they are

1. in AS3-package `joo`
2. are `[Native]`
3. with _no_ `required` attribute

Then, they lead to `joo` being imported.

fwienber replied:
What about `KeyEvent`, does it really belong in this list? `js.KeyEvent` is a very special case below...

aholtkamp replied:
Agreed.

Forgot to remove KeyEvent from that list.

fwienber commented on jangaroo/jangaroo-compiler/src/main/java/net/jangaroo/jooc/backend/TypeScriptCodeGenerator.java:
```diff
@@ -101,6 +101,41 @@
+  );
+  private static final List<String> JANGAROO_RUNTIME_REMOVED_NAMES = Arrays.asList(
+          "AS3.Error",
+          "joo.baseUrl",
+          "joo.boundMethod",
+          "joo.classLoader",
+          "joo.compilerVersion",
+          "joo.DynamicClassLoader",
+          "joo.getOrCreatePackage",
+          "joo.getRelativeClassUrl",
+          "joo.JavaScriptObject",
+          "joo.ClassDeclaration",
+          "joo.loadDebugScript",
+          "joo.loadModule",
+          "joo.loadScript",
+          "joo.loadScriptAsync",
+          "joo.MemberDeclaration",
+          "joo.NativeClassDeclaration",
+          "joo.require",
+          "joo.resolveUrl",
+          "joo.runtimeApiVersion",
+          "joo.StandardClassLoader",
+          "joo.SystemClassLoader"
```
To avoid this list, too, we _could_ add `[Deprecated(since="4.1", replacement="none")]` or the like to all these Compilation Units and check that.

aholtkamp replied:
added [Deprecated] annotation in jangaroo-libs (see https://github.com/CoreMedia/jangaroo-libs/commit/3b525c929651802b030b8c1ed399c5ed5c4e18e9)

fwienber commented on jangaroo/jangaroo-compiler/src/main/java/net/jangaroo/jooc/backend/TypeScriptCodeGenerator.java:
```diff
@@ -280,16 +315,33 @@ public void visitCompilationUnit(CompilationUnit compilationUnit) throws IOExcep
+      }
+      if (JANGAROO_RUNTIME_NATIVE_NAMES_LEADING_TO_JOO_IMPORT.stream().anyMatch(
+              noLongerNativeName -> compilationUnit.getCompileDependencies().contains(noLongerNativeName)
+      )) {
```
See my comment above.

fwienber committed:
[TS] Adapt expected ITest code to jangaroo-runtime name / import changes

ITests are green again.

fwienber committed:
[TS] Rule-based solution to detect when to import joo

Since all [Native] compilation units in package "joo" without
a 'require' attribute lead to the import of package
@jangaroo/runtime/joo, we can replace the long list of
compilation unit names by this rule.

aholtkamp committed:
consider [Deprecated replacement="none"] for "joo" instead of hard wired list

aholtkamp committed:
replace "ResourceManager.getInstance()" with "resourceManager"

fwienber committed:
[TS] Fix: Remove duplicate dependent CUs after mapping 'resourceManager'

Mapping 'mx.resources.ResourceManager' to 'resourceManager' may
lead to a duplicate CompilationUnit 'resourceManager'. This just
by coincidence did not cause a duplicate import, because
'resourceManager' is in the top-level package, so its fully-qualified
name is also 'resourceManager', and same-name imports are filtered
out later.
It is much better to remove duplicate CUs after mapping 'resourceManager',
keeping the original import order by using a LinkedHashSet.

fwienber committed:
[TS] Get rid of KeyEvent as type

If used as a type, (js.)KeyEvent is now replaced by
browser built-in, TypeScript lib-dom KeyboardEvent.
The only usages of KeyEvent that remain are when
accessing the (legacy) constants.
